apps/glusterfs/device_entry.go

```diff
@@ -573,30 +573,40 @@ func markDeviceFailed(db wdb.DB, id string, force bool) error {
 func (d *DeviceEntry) DeleteBricksWithEmptyPath(tx *bolt.Tx) error {
 	godbc.Require(tx != nil)
 	var bricksToDelete []*BrickEntry
+	var bricksToRemoveFromDevice []string
 
 	logger.Debug("Deleting bricks with empty path on device [%v].",
 		d.Info.Id)
 
 	for _, id := range d.Bricks {
 		brick, err := NewBrickEntryFromId(tx, id)
 		if err == ErrNotFound {
-			logger.Warning("Ignoring nonexistent brick [%v] on "+
-				"disk [%d].", id, d.Info.Id)
-			continue
-		}
-		if err != nil {
+			bricksToRemoveFromDevice =
+				append(bricksToRemoveFromDevice, id)
+		} else if err != nil {
 			logger.LogError("Unable to fetch brick [%v] from db: %v",
 				id, err)
 			return err
-		}
-		if brick.Info.Path == "" {
+		} else if brick.Info.Path == "" {
 			bricksToDelete = append(bricksToDelete, brick)
 		}
 	}
+
+	for _, id := range bricksToRemoveFromDevice {
+		logger.Debug("Removing unreferenced brick [%v] from device [%v].",
+			id, d.Info.Id)
+		d.BrickDelete(id)
+	}
+	err := d.Save(tx)
+	if err != nil {
+		logger.LogError("Unable to save device %v: %v", d.Info.Id, err)
+		return err
+	}
+
```

Comment:
Move this section below the bricksToDelete section. That way we won't leave any orphans bricks again.

If you agree with the approach but don't have time to, I would be happy to do the update.

Reply:
I talked this over with @raghavendra-talur  and we both agree that changes need to be made here so that we don't delete the brick entries but leave the references on the device. One addtional suggestion I have is to avoid looping twice and instead create a new list of "good bricks" as you go. Then after the end of the loop assign the new good-brick list to d.Bricks.

Reply:
@raghavendra-talur writes:
> Move this section below the bricksToDelete section. That way we won't leave any orphans bricks again.

I don't fully understand yet: How would we possibly end up leaving (new) orphan bricks again here, due to this brick from device removal being above the loop to remove the bricks without path? That loop also removes the removed bricks-with-empty-path from the referencing device.

Reply:
@phlogistonjohn writes:
>  I talked this over with @raghavendra-talur and we both agree that changes need to be made here so that we don't delete the brick entries but leave the references on the device.

I don't get the problem yet: In the loop over the collected bricks with empty path, we do

```
   d.BrickDelete(brick.Info.Id)
   d.Save(tx)
   ...
```
so that leaves no new references to nonexisting bricks.

>  One additional suggestion I have is to avoid looping twice and instead create a new list of "good bricks" as you go. Then after the end of the loop assign the new good-brick list to d.Bricks.

I am giving it a try, but I have to say that I find the current proposal much more natural: It is removing the brick ids that reference a nonexisting brick in one loop and then removes the brick-ids that reference bricks with empty path when it actually removes the brick entry itself. That seems like a natural grouping to me even though the multiple loops may not be highly optimized. But I am not sure that optimization is very important here, since this should not need to be a highly performant operation.

If I prepare the complete new list of bricks for the device in the first loop and just assign it over the device's brick list afterwards, I have to work a little harder in the first loop (because of the sorted string list type), and spare the loop over the (few?) brick IDs that don't reference an existing brick. The end result will be exactly the same due to all this being wrapped in a transaction.

I will do the corresponding change and look at the result, but the assignment of the device's brick list instead of operating on it makes me feel slightly uneasy.